subprojects/language-scala/src/main/java/org/gradle/api/internal/tasks/scala/ZincScalaCompilerFactory.java

```diff
@@ -43,15 +52,107 @@
 import java.util.Arrays;
 import java.util.Collections;
 import java.util.List;
+import java.util.Objects;
 import java.util.Properties;
+import java.util.Set;
+import java.util.concurrent.Callable;
+import java.util.stream.Collectors;
 
 import static org.gradle.cache.internal.filelock.LockOptionsBuilder.mode;
 
 public class ZincScalaCompilerFactory {
     private static final Logger LOGGER = Logging.getLogger(ZincScalaCompilerFactory.class);
+    private static final int CLASSLOADER_CACHE_SIZE = 4;
+    private static final int COMPILER_CLASSLOADER_CACHE_SIZE = 4;
+    private static final GuavaBackedClassLoaderCache<HashCode> CLASSLOADER_CACHE = new GuavaBackedClassLoaderCache<HashCode>(CLASSLOADER_CACHE_SIZE);
+    private static final TimeCheckingClassLoaderCache COMPILER_CLASSLOADER_CACHE = new TimeCheckingClassLoaderCache(COMPILER_CLASSLOADER_CACHE_SIZE);
 
-    static ZincScalaCompiler getCompiler(CacheRepository cacheRepository, final Iterable<File> scalaClasspath) {
-        ScalaInstance scalaInstance = getScalaInstance(scalaClasspath);
+    static class TimestampedFile {
+        private final File file;
+        private final long timestamp;
+
+        public TimestampedFile(File file) {
+            this.file = file;
+            this.timestamp = IO.getModifiedTimeOrZero(file);
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) {
+                return true;
+            }
+            if (o == null || getClass() != o.getClass()) {
+                return false;
+            }
+            TimestampedFile that = (TimestampedFile) o;
+            return timestamp == that.timestamp &&
+                Objects.equals(file, that.file);
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(file, timestamp);
+        }
+    }
+
+    /**
+     * This class implements AbstractClassLoaderCache in a way that allows safe
+     * resource release when entries are evicted.
+     *
+     * Cache is based on file timestamps, because this method is used in sbt implementation.
+     */
+    static class TimeCheckingClassLoaderCache implements AbstractClassLoaderCache {
```

Comment:
The `AbstractClassLoaderCache` does not exist in Zinc 1.2.0 which is supported by Gradle.